### PR TITLE
Finalize the integration of the TF-PSA-Crypto build system into the Mbed TLS one

### LINKED ISSUE #9615: Finalize the integration of the TF-PSA-Crypto build system into the Mbed TLS one
### Suggested enhancement
Follow-up of #9283/#9445, documentation, packaging and installation of TF-PSA-Crypto. The CMakeLists.txt in tf-psa-crypto is just the top level cmake file of TF-PSA-Crypto (not the case in #9445).

Depends on: #9283 

### Justification
Mbed TLS needs this for the repo split.

### PR DESCRIPTION
**Pull request by ronald-cron-arm**
## Description
Fixes #9615 

## PR checklist
- [x] **changelog** not required
- [x] **development PR** provided here
- [x] **framework PR** not required
- [x] **3.6 PR** not required because: repo split specific work 
- [x] **2.28 PR** not required because: repo split specific work
- **tests**  provided: one additional test program, cmake_subproject

### CODE REVIEW

File: programs/test/dlopen.c [hunk 1]
```diff
@@ -73,8 +74,16 @@ int main(void)
 #endif  /* MBEDTLS_X509_CRT_PARSE_C */
 
 #if defined(MBEDTLS_MD_C)
-    void *crypto_so = dlopen(CRYPTO_SO_FILENAME, RTLD_NOW);
-    CHECK_DLERROR("dlopen", CRYPTO_SO_FILENAME);
+    const char *crypto_so_filename = NULL;
+    void *crypto_so = dlopen(MBEDCRYPTO_SO_FILENAME, RTLD_NOW);
+    if (dlerror() == NULL) {
```
Comment (minosgalanakis) [inline programs/test/dlopen.c:79]: Why is that required? This feels like it is fixing a programs's bug?

**Review by minosgalanakis — COMMENTED** [review]
Looks good overall on the first pass. Some minor comments/questions.

File: tf-psa-crypto/CMakeLists.txt [hunk 2]
```diff
@@ -31,43 +37,491 @@ cmake_policy(SET CMP0011 NEW)
...
+                         "${CTR_DRBG_128_BIT_KEY_WARN_L3}"
+                         "${WARNING_BORDER}")
+
+# Python 3 is only needed here to check for configuration warnings.
+if(NOT CMAKE_VERSION VERSION_LESS 3.15.0)
+    set(Python3_FIND_STRATEGY LOCATION)
+    find_package(Python3 COMPONENTS Interpreter)
+    if(Python3_Interpreter_FOUND)
+        set(TF_PSA_CRYPTO_PYTHON_EXECUTABLE ${Python3_EXECUTABLE})
+    endif()
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply tf-psa-crypto/CMakeLists.txt:146]: > Should we fail `CMAKE_VERSION ` < `3.15.1` and no Python is discovered? If it would not reach this point of the configuration without an interpreter, the check may be redudant.
With a release tree that includes the generated files we do not want to fail I think.

> Also why are we not using the `TF_PSA_CRYPTO_PYTHON_EXECUTABLE`?
It is used just below and in various places in tf-psa-crypto CMake build system thus I am not sure what you mean here.

File: tf-psa-crypto/CMakeLists.txt [hunk 3]
```diff
@@ -31,43 +37,491 @@ cmake_policy(SET CMP0011 NEW)
...
+    # Old GCC versions hit a performance problem with test_suite_pkwrite
+    # "Private keey write check EC" tests when building with Asan+UBSan
+    # and -O3: those tests take more than 100x time than normal, with
+    # test_suite_pkwrite taking >3h on the CI. Observed with GCC 5.4 on
+    # Ubuntu 16.04 x86_64 and GCC 6.5 on Ubuntu 18.04 x86_64.
+    # GCC 7.5 and above on Ubuntu 18.04 appear fine.
+    # To avoid the performance problem, we use -O2 when GCC version is lower than 7.0.
+    # It doesn't slow down much even with modern compiler versions.
+    target_compile_options(${target} PRIVATE $<$<CONFIG:ASan>:-fsanitize=address -fno-common -fsanitize=undefined -fno-sanitize-recover=all>)
+    if (GCC_VERSION VERSION_LESS 7.0)
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply tf-psa-crypto/CMakeLists.txt:285]: A log would be invasive in the cmake configuration logs as this code is called for each target. Furthermore an ASan build is a test build thus this seems okay to me.

File: programs/test/dlopen.c (hunk 1 above)
Comment (ronald-cron-arm) [reply programs/test/dlopen.c:79]: It is required because the crypto library could be named libmbedcrypto or libtfpsacrypto depending the context the program is executed in. Thus we try both. Getting rid of libmbedcrypto comes with significant work thus I'd prefer to do it later.

**Review by minosgalanakis — COMMENTED** [review]
LGTM.

The conflict is a minor-one because of the introduction of a new component. 

Tested locally the following:
- Build as part of MbedTLS
- Build standalone
- Build as part of MbedTls out of source.

Will approve after rebase.

**Review by valeriosetti — DISMISSED** [review]
Approving to mark that this PR is OK to me. I will reapprove once the conflict is resolved.

**Comment by ronald-cron-arm** [comment]
While fixing the everest issue, I thought about a simplification. Added an extra commit (last one) for that.

**Comment by ronald-cron-arm** [comment]
I've rebased to resolve the conflict and added three additional commits. CI green, @minosgalanakis @valeriosetti please have another look.

File: tf-psa-crypto/CMakeLists.txt [hunk 4]
```diff
@@ -31,43 +37,492 @@ cmake_policy(SET CMP0011 NEW)
...
+
+    if(MBEDTLS_FATAL_WARNINGS)
+        target_compile_options(${target} PRIVATE /WX)
+    endif(MBEDTLS_FATAL_WARNINGS)
+endfunction(set_msvc_base_compile_options)
+
+function(set_config_files_compile_definitions target)
+    # Pass-through MBEDTLS_CONFIG_FILE, MBEDTLS_USER_CONFIG_FILE,
+    # TF_PSA_CRYPTO_CONFIG_FILE and TF_PSA_CRYPTO_USER_CONFIG_FILE
+    if(MBEDTLS_CONFIG_FILE)
```
Comment (valeriosetti) [inline tf-psa-crypto/CMakeLists.txt:352]: Nothing to change, just my curiosity here: are we using `MBEDTLS_CONFIG_FILE` and `MBEDTLS_USER_CONFIG_FILE` also in the TF-psa-crypto repo? I thought it was self contained which, in my mind, it means also from the configuration point of view, but probably I'm missing something

**Review by valeriosetti — COMMENTED** [review]
Looking at this again I found some minor questions. If you think they are unrelevant, I can just approve as-is ;)

File: tf-psa-crypto/CMakeLists.txt (hunk 4 above)
Comment (valeriosetti) [reply tf-psa-crypto/CMakeLists.txt:352]: Yeah, I thought that could be the case. Well, since you confirmed that, I'm OK with this and I can resolve this comment. Thanks

File: CMakeLists.txt [hunk 5]
```diff
@@ -348,8 +369,30 @@ add_subdirectory(framework)
 
 add_subdirectory(include)
 
+set(TF_PSA_CRYPTO_TARGET_PREFIX ${MBEDTLS_TARGET_PREFIX} CACHE STRING "")
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply CMakeLists.txt:372]: Yes it fits better here.

File: tf-psa-crypto/CMakeLists.txt (hunk 4 above)
Comment (ronald-cron-arm) [reply tf-psa-crypto/CMakeLists.txt:352]: We will remove those when #9567 is completed.

**Review by valeriosetti — APPROVED** [review]
Thanks for addressing my comments. LGTM :)

**Comment by ronald-cron-arm** [comment]
> Thanks for addressing my comments. LGTM :)

Thanks for your review and comments.

**Review by minosgalanakis — APPROVED** [review]
LGTM
